src/main/kotlin/org/rust/ide/docs/RustDocumentationProvider.kt

```diff
@@ -60,15 +60,35 @@ class RustDocumentationProvider : AbstractDocumentationProvider() {
         }
 
         val signatureStart = signatureStartElement?.startOffsetInParent ?: 0
-        val signatureEnd = block?.startOffsetInParent ?: textLength
+
+        // pick last of: where clause, return type, or closing brace of the parameters
+        // if all else fails, drop down to the length of the current element
+        val signatureEnd = if (block != null)
+            block!!.startOffsetInParent
+        else if (whereClause != null)
+            whereClause!!.let { it.startOffsetInParent + it.textLength }
+        else if (retType != null)
+            retType!!.let { it.startOffsetInParent + it.textLength }
+        else if (parameters != null)
+            parameters!!.let { it.startOffsetInParent + it.textLength }
+        else
+            textLength
 
         val identStart = identifier.startOffsetInParent
         val identEnd = identStart + identifier.textLength
         check(signatureStart <= identStart && identStart <= identEnd &&
             identEnd <= signatureEnd && signatureEnd <= textLength)
 
         val beforeIdent = text.substring(signatureStart, identStart).escaped
-        val afterIdent = text.substring(identEnd, signatureEnd).trimEnd().escaped
+        val afterIdent = text.substring(identEnd, signatureEnd)
+                .replace(Regex("""\n"""), "")    // collapse signature into a single line
+                .replace(Regex("""\s\s+"""), "") // collapse consecutive whitespace runs into a single whitespace
+                .replace(Regex("""\(\s"""), "(") // remove whitespace after a parens
+                .replace(Regex("""\s\)"""), ")") // remove whitespace before a parens
+                .replace(Regex(""",\s"""), ",")  // normalize all commas to ?,?
+                .replace(Regex(""","""), ", ")   // now that commas don't have following spaces, add one
+                .trimEnd()                       // finally, trim any ending whitespace
```

Comment:
>would be transformed to

It would if we `\s+ -> \s` before

>Running formatter on this would be ideal

Yep, though we would need to collapse newlines anyway.